crates/runix/src/uri_parser.rs

```diff
@@ -0,0 +1,971 @@
+            return locked_ref.map(|_| unreachable!());
+        }
+        Ok(LockedFlakeRef {
+            input,
+            original_ref: original_ref.unwrap(), // already checked for error case
+            resolved_ref: resolved_ref.unwrap(), // already checked for error case
+            locked_ref: locked_ref.unwrap(),     // already checked for error case
+        })
+    }
+}
+
+/// A flake reference URI pointing to an installable that has been resolved by Nix
+#[derive(Debug, Clone)]
+pub struct InstallableFlakeRef {
+    pub input: Input,
+    pub attr_path: Vec<String>,
+    pub outputs: InstallableOutputs,
+    pub r#ref: ParsedFlakeReference,
+}
+
+impl TryFrom<GenericParsedURI> for InstallableFlakeRef {
+    type Error = UriParseError;
+
+    fn try_from(parsed: GenericParsedURI) -> Result<Self, Self::Error> {
+        let Some(input) = parsed.input else {
+            return Err(UriParseError::MissingAttribute("input".to_string()));
+        };
+        let Some(attr_path) = parsed.attr_path else {
+            return Err(UriParseError::MissingAttribute("attrPath".to_string()));
+        };
+        let Some(outputs) = parsed.outputs else {
+            return Err(UriParseError::MissingAttribute("outputs".to_string()));
+        };
+        // Calling this 'rref' so we don't have to keep using raw identifier syntax
+        let Some(rref) = parsed.r#ref else {
+            return Err(UriParseError::MissingAttribute("ref".to_string()));
+        };
+        let rref = ParsedFlakeReference::try_from(rref);
+        if rref.is_err() {
+            return rref.map(|_| unreachable!());
+        }
+        Ok(InstallableFlakeRef {
+            input,
+            attr_path,
+            outputs,
+            r#ref: rref.unwrap(), // already checked for error case
+        })
+    }
+}
+
+/// A completely generic parsed URI i.e. a type that can represent a call to `parser_util`
+/// with any of the flags
+#[derive(Debug, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+struct GenericParsedURI {
+    #[serde(deserialize_with = "deserialize_input")]
+    input: Option<Input>,
+    original_ref: Option<DeserializedFlakeRef>,
+    resolved_ref: Option<DeserializedFlakeRef>,
+    locked_ref: Option<DeserializedFlakeRef>,
+    r#ref: Option<DeserializedFlakeRef>,
+    attr_path: Option<Vec<String>>,
+    outputs: Option<InstallableOutputs>,
+    authority: Option<Option<String>>,
+    base: Option<String>,
+    fragment: Option<String>,
+    path: Option<String>,
+    query: Option<HashMap<String, Option<String>>>,
+    scheme: Option<Scheme>,
+}
+
+/// A type representing a Nix-parsed URL scheme
+#[derive(Debug, Clone, Deserialize)]
+pub struct Scheme {
+    pub application: Option<String>,
+    pub full: String,
+    pub transport: String,
+}
+
+/// The type and contents of the input the URI parser received.
+#[derive(Debug, Deserialize, Clone)]
+pub enum Input {
+    String(String),
```

Comment:
When is this just a string?

Reply:
It's almost always just a string. It's not a string when the user supplies attributes like
```
{
  type = "github";
  owner = "NixOS";
  repo = "nixpkgs";
}
```

Reply:
I've changed this to 
```
pub enum Input {
    URL(String),
    Attrs(Attrs),
}
```
with some comments for clarity